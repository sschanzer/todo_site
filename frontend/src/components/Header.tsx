import React from "react";
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Form from "react-bootstrap/Form";
import { Button, InputGroup } from "react-bootstrap";
import { ITask } from "../App";
import { useEffect, useState } from "react";
import axios from "axios";

// set up to pass pending task and setpending tasks into Header
interface HeaderProps {
  allTasks: ITask[];
  setAllTasks: (allTasks: ITask[]) => void;
}

// export so we can run tests on createTaskResponse
export interface createTaskResponse {
  itemCreated: boolean;
  id: number;
}

export const createTask = async (
  taskTitle: string
): Promise<createTaskResponse> => {
  let response = await axios.post("new_task/", {
    name: taskTitle,
  });
  return response["data"];
};

export const isTaskTitleEmpty = (taskTile: string) => {
  let cleanInput = taskTile.replaceAll(" ", "");
  if (cleanInput.length >= 1 && cleanInput != "") {
    return false;
  } else {
    return true;
  }
};

export const Header: React.FC<HeaderProps> = ({ allTasks, setAllTasks }) => {
  const [newTask, setNewTask] = useState("");
  const [isSubmittedDisabled, setIsSubmittedDisabled] = useState(true);

  useEffect(() => {
    setIsSubmittedDisabled(isTaskTitleEmpty(newTask));
  }, [newTask]);

  const createNewTask = async (
    task: string,
    event?: React.FormEvent<HTMLFormElement>
  ) => {
    event?.preventDefault();
    let response = await createTask(task);
    console.log(response);
    if (response.itemCreated) {
      setAllTasks([
        ...allTasks,
        { id: response.id, title: task, completed: false },
      ]);
      setNewTask("");
    }
  };

  return (
    <Container>
      <Row>
        <Col xs={4}></Col>
        <Col xs={8}>
          <Form
            style={{ position: "relative" }}
            onSubmit={(event) => createNewTask(newTask, event)}
          >
            <Form.Control
              value={newTask}
              id="createTaskInput"
              onChange={(event) => setNewTask(event.target.value)}
            />
            <Button
              variant="success"
              id="createTaskButton"
              disabled={isSubmittedDisabled}
              type="submit"
            >
              +
            </Button>
          </Form>
        </Col>
      </Row>
    </Container>
  );
};

export default Header;
