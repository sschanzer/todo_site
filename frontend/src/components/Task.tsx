import React from "react";
import { ITask } from "../App";
import Form from "react-bootstrap/Form";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import axios from "axios";
import { Button } from "react-bootstrap";

export interface TaskProps {
  task: ITask;
  selectedTasks: number[];
  setSelectedTasks: (selectedTasks: number[]) => void;
  allTasks: ITask[];
  setAllTasks: (allTasks: ITask[]) => void;
}

export const changeTaskStatus = async (id: number) => {
  let response = await axios.put(`task/${id}/`);
  return response.data.changed;
};

export const deleteTask = async (id: number) => {
  let response = await axios.delete(`task/${id}/`);
  return response.data.success;
};

export const Task: React.FC<TaskProps> = ({
  task,
  allTasks,
  setAllTasks,
  selectedTasks,
  setSelectedTasks,
}) => {
  const changeStatus = async (clicked: boolean, taskToChange: ITask) => {
    let response = await changeTaskStatus(taskToChange["id"]);
    if (response) {
      setAllTasks(allTasks.filter((task) => task != taskToChange));
      taskToChange.completed = clicked;
      setAllTasks([...allTasks]);
    }
  };

  const deleteATask = async (id: number) => {
    let response = await deleteTask(id);
    if (response) {
      setAllTasks(allTasks.filter((task) => task.id !== id));
    }
  };

  return (
    <Row className="task">
      <Col>
        <Form.Check
          id={`taskSelectedBtn${task.id}`}
          type="checkbox"
          onChange={(event) =>
            event.target.checked
              ? setSelectedTasks([...selectedTasks, task.id])
              : setSelectedTasks(selectedTasks.filter((id) => id !== task.id))
          }
        />
      </Col>
      <Col xs={8} id={`task${task.id}`} className="taskTitle">
        {task.title}
      </Col>
      <Col className="checkHolder" xs={2}>
        <Form.Check
          id={`taskCheck${task.id}`}
          type="checkbox"
          className="check"
          checked={task.completed}
          onChange={(event) => changeStatus(event.target.checked, task)}
        />
        <Form.Label for="checkbox">
          {task.completed ? "done" : "pending"}
        </Form.Label>
        <Col xs={1}>
          <Button
            id={`deleteBtn$(task.id)`}
            className="delBtn"
            variant="danger"
            onClick={() => deleteATask(task.id)}
          >
            Delete
          </Button>
        </Col>
      </Col>
    </Row>
  );
};
