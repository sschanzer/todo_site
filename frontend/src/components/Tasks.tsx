import React from "react";
import { myTasks } from "../App";
import Form from "react-bootstrap/Form";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";

interface IdProps {
  task: myTasks;
  selected: number[];
  setSelected: (selected: number[]) => void;
}

export const Tasks: React.FC<IdProps> = ({ task, selected, setSelected }) => {
  return (
    <div className="task">
      <Row>
        <Col xs={8} style={{ display: "flex" }}>
          <p>{task.title}</p>
        </Col>
        <Col xs={4} style={{ display: "flex" }}>
          <Form.Check type="checkbox" label="select" />
        </Col>
      </Row>
    </div>
  );
};
