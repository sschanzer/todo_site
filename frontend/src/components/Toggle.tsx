import ToggleButton from "react-bootstrap/ToggleButton";
import ToggleButtonGroup from "react-bootstrap/ToggleButtonGroup";

interface ToggleProps {
  showPending: boolean;
  setShowPending: (showPending: boolean) => void;
  showCompleted: boolean;
  setShowCompleted: (showCompleted: boolean) => void;
}

export const Toggle: React.FC<ToggleProps> = ({
  showPending,
  setShowPending,
  showCompleted,
  setShowCompleted,
}) => {
  return (
    <ToggleButtonGroup defaultValue={[1, 3]} size="sm" type="checkbox">
      <ToggleButton
        variant="outline-primary"
        id="tbg-btn-1"
        checked={showPending}
        defaultChecked={showPending}
        onChange={(e) =>
          showCompleted
            ? setShowPending(e.target.checked)
            : [
                setShowPending(e.target.checked),
                setShowCompleted(!e.target.checked),
              ]
        }
        value={1}
      >
        Pending
      </ToggleButton>
      <ToggleButton
        variant="outline-primary"
        id="tbg-btn-3"
        checked={showCompleted}
        defaultChecked={showCompleted}
        onChange={(e) =>
          showPending
            ? setShowCompleted(e.target.checked)
            : [
                setShowCompleted(e.target.checked),
                setShowPending(!e.target.checked),
              ]
        }
        value={3}
      >
        Complete
      </ToggleButton>
    </ToggleButtonGroup>
  );
};
