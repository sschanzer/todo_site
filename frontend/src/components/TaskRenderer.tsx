import { ITask } from "../App";
import { Task, TaskProps } from "./Task";

interface TaskRendererProps {
  completed: boolean;
  allTasks: ITask[];
  setAllTasks: (pendTasks: ITask[]) => void;
}
export const TaskRenderer: React.FC<TaskRendererProps> = ({
  setAllTasks,
  completed,
  allTasks,
}) => {
  return (
    <>
      {allTasks.map((task) =>
        task.completed === completed ? (
          <Task allTasks={allTasks} setAllTasks={setAllTasks} task={task} />
        ) : null
      )}
    </>
  );
};
