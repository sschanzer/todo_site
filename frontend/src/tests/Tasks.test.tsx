import { Tasks } from "../components/Tasks";
import { describe, expect, it } from "vitest";
import TestRenderer from "react-test-renderer";


describe("take a snapshot of Tasks component", () => {
    it("gives a snapshot of Tasks", () => {
        let selected:number[] = []
        const setSelected = (v:number[]) => {selected = v}
        const header = TestRenderer.create(<Tasks setSelected={setSelected} selected={selected} task={{"id":1, "title":"Create Test", "completed":true}}/>)
        expect(header).toMatchSnapshot()
    })
})