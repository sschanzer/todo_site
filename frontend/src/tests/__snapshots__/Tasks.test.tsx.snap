// Vitest Snapshot v1

<<<<<<< HEAD
exports[`Tasks > snapshot of Tasks component > gives a snapshot of Tasks 1`] = `
=======
exports[`Tasks > Take a snapshot of Tasks component > returns a snapshot of Tasks 1`] = `
<div
  className="task"
>
  <div
    className="row"
  >
    <div
      className="col-8"
      style={
        {
          "display": "flex",
        }
      }
    >
      <p>
        Create Test
      </p>
    </div>
    <div
      className="col-4"
      style={
        {
          "display": "flex",
        }
      }
    >
      <div
        className="form-check"
      >
        <input
          className="form-check-input"
          disabled={false}
          type="checkbox"
        />
        <label
          className="form-check-label"
          title=""
        >
          select
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`take a snapshot of Tasks component > gives a snapshot of Tasks 1`] = `
>>>>>>> TODO-001
<div
  className="task"
>
  <div
    className="row"
  >
    <div
      className="col-8"
      style={
        {
          "display": "flex",
        }
      }
    >
      <p>
        Create Test
      </p>
    </div>
    <div
      className="col-4"
      style={
        {
          "display": "flex",
        }
      }
    >
      <div
        className="form-check"
      >
        <input
          className="form-check-input"
          disabled={false}
          type="checkbox"
        />
        <label
          className="form-check-label"
          title=""
        >
          select
        </label>
      </div>
    </div>
  </div>
</div>
`;
